File: src/main/java/ru/frostman/scalable/reactor/ExtSelector.java
package ru.frostman.scalable.reactor;

import ru.frostman.scalable.reactor.handlers.AcceptHandler;
import ru.frostman.scalable.reactor.handlers.ConnectHandler;
import ru.frostman.scalable.reactor.handlers.SelectorAttachment;
import ru.frostman.scalable.reactor.utils.EventExecutor;
import ru.frostman.scalable.reactor.utils.impl.StaticThreadEventExecutor;

import java.io.IOException;
import java.nio.channels.SelectionKey;
import java.nio.channels.Selector;
import java.util.Iterator;
import java.util.Set;

/**
 * This class is I/O Event queue for selector. It provides high level
 * asynchronized methods to manage all operations on the selector, like
 * registering and unregistering channels, or updating the events of
 * interest for each monitored socket.
 *
 * It supports pending invocation of methods. Only thread of ExtSelector
 * can access selector object and invoke methods on it or access all
 * sockets managed by it.
 *
 * @author slukjanov aka Frostman
 */

public class ExtSelector implements Runnable {
    /**
     * Logging handler.
     */
  //  private static final Logger log = Logger.getLogger(ExtSelector.class);

    /**
     * Selector used for I/O working.
     */
    private final Selector selector;

    /**
     * Flag telling that selector must work
     * or gracefully terminated.
     */
    private boolean work = false;

    /**
     * Executor for processing handlers.
     */
    private final EventExecutor executor;

    /**
     * Creates a new selector and the associated thread started by this
     * constructor.
     *
     * @param threads number of threads for executor
     *
     * @throws IOException iff Selector.open() crushed.
     */
    public ExtSelector(int threads) throws IOException {
        executor = new StaticThreadEventExecutor(threads, 300000);
        selector = Selector.open();
    }

    /**
     * Starts main loop.
     */
    public void start() {
        if (work) {
            throw new IllegalStateException();
        }

        work = true;
        //log.info("Server starts successfully");
        run();
    }

    /**
     * Changes work flag to false. All pending tasks will be executed
     * before dying.
     */
    public void stop() {
        work = false;
        selector.wakeup();
    }

    /**
     * Main loop. This is where i/o events dispatching and executes
     * pending
     */
    @Override
    public void run() {
        try {
            while (true) {
                if (!work) {
                    break;
                }

                int selectedKeys;
                try {
                    selectedKeys = selector.select();
                } catch (IOException ioe) {
                    ioe.printStackTrace();
                    continue;
                }

                if (selectedKeys == 0) {
                    // here we can check idle connections
                    continue;
                }

                Iterator<SelectionKey> it = selector.selectedKeys().iterator();
                while (it.hasNext()) {
                    SelectionKey sk = it.next();
                    it.remove();
                    try {
                        SelectorAttachment attachment = (SelectorAttachment) sk.attachment();

                        if (sk.isAcceptable()) {
                            ((AcceptHandler) attachment).doAccept();
                        } else if (sk.isConnectable()) {
                            ((ConnectHandler) attachment).doConnect();
                        } else {
                            ConnectionHandler connection = (ConnectionHandler) attachment;
                            if (sk.isValid() && sk.isReadable() && connection.removeReadInterest()) {
                                executor.execute(connection.getReadEvent());
                            }

                            if (sk.isValid() && sk.isWritable() && connection.removeWriteInterest()) {
                                executor.execute(connection.getWriteEvent());
                            }
                        }
                    } catch (Exception e) {
                        log.error("Exception in processing SelectionKey", e);
                    }
                }
            }
            log.info("Server stop");
        } catch (Throwable t) {
            closeSelector();

            log.error("Throwable in main selector loop", t);
        }
    }

    /**
     * Safely close all selection keys and selector.
     */
    private void closeSelector() {
        Set<SelectionKey> keys = selector.keys();
        for (SelectionKey key : keys) {
            try {
                key.channel().close();
            } catch (IOException e) {
                // no operation
            }
        }
        try {
            selector.close();
        } catch (IOException e) {
            // no operation
        }
    }

    public Selector getSelector() {
        return selector;
    }

    public void wakeup() {
        //TODO check it, if select(timeout)
        selector.wakeup();
    }
}

File: src/reactor/ExtSelector.kt
package reactor

/** 
 * @author slukjanov aka Frostman
 */

import java.nio.channels.Selector
import java.util.concurrent.Executors
import java.nio.channels.ServerSocketChannel
import java.net.InetSocketAddress
import java.nio.channels.SelectableChannel
import java.nio.channels.SelectionKey
import java.nio.channels.SocketChannel
import java.util.concurrent.atomic.AtomicBoolean

class ExtSelector(val workersCount : Int, val connHandlers : ConnectionHandlersStrategy) : Thread() {
    private val selector = Selector.open().sure()
    private var work = true
    private val executor = Executors.newFixedThreadPool(workersCount).sure()

    override fun run() {
        connHandlers.init(selector)

        while (work) {
            val selectedKeys = selector.select()
            if (selectedKeys == 0) continue

            val it = selector.selectedKeys().sure().iterator().sure()
            while (it.hasNext()) {
                val sk = it.next().sure()
                it.remove()

                val obj = sk.attachment().sure()
                val attachment : ConnectionHandler = if (obj is ConnectionHandler) obj else throw IllegalStateException()

                if (sk.isAcceptable()) {
                    attachment.doAccept()
                } else if (sk.isConnectable()) {
                    attachment.doConnect()
                } else {
                    if (sk.isValid() && sk.isReadable() && attachment.remReadInterest()) {
                        executor.execute(attachment.readEvent)
                    }

                    if (sk.isValid() && sk.isWritable() && attachment.remWriteInterest()) {
                        executor.execute(attachment.writeEvent)
                    }
                }
            }
        }
    }
}

abstract class ConnectionHandlersStrategy(val host : String, val port : Int) {
    abstract fun init(val selector : Selector) : Unit

    abstract fun create(val selector : Selector) : ConnectionHandler
}

abstract class ConnectionHandler(var selectionKey : SelectionKey? = null, val selector : Selector) {
    val readInterest = AtomicBoolean(false)
    val writeInterest = AtomicBoolean(false)

    abstract val readEvent : Runnable
    abstract val writeEvent : Runnable

    abstract fun doAccept()
    abstract fun doConnect()

    fun addReadInterest() = readInterest.compareAndSet(false, true)
    fun addWriteInterest() = writeInterest.compareAndSet(false, true)
    fun remReadInterest() = readInterest.compareAndSet(true, false)
    fun remWriteInterest() = writeInterest.compareAndSet(true, false)
}

class ServerConnectionHandlersStrategy(host : String, port : Int) : ConnectionHandlersStrategy(host, port) {
    val serverChannel = ServerSocketChannel.open().sure()
    var selector : Selector? = null

    override fun init(val selector : Selector) {
        this.selector = selector
        serverChannel.socket().sure().bind(InetSocketAddress(host, port))
        serverChannel.registerChannel(selector, SelectionKey.OP_ACCEPT, create(selector))
    }

    override fun create(val selector : Selector) = ServerConnectionHandler(selector.sure(), serverChannel)
}

class ServerConnectionHandler(selector : Selector, val serverChannel : ServerSocketChannel) : ConnectionHandler(null, selector){
    private var socketChannel : SocketChannel? = null

    override fun doAccept() {
        socketChannel = serverChannel.accept().sure()
        socketChannel?.registerChannel(selector, SelectionKey.OP_READ, this)
    }

    override fun doConnect() {
        throw UnsupportedOperationException()
    }

    override val readEvent = object : Runnable {
        override fun run() {
        }
    }

    override val writeEvent = object : Runnable {
        override fun run() {
        }
    }
}

fun SelectableChannel.registerChannel(selector : Selector, selectionKeys : Int, handler : ConnectionHandler) {
    var sk : SelectionKey
    if (this.isRegistered()) {
        sk = this.keyFor(selector).sure()
        sk.interestOps(selectionKeys)
        sk.attach(handler)
    } else {
        this.configureBlocking(false)
        sk = this.register(selector, selectionKeys, handler).sure()
    }
    handler.selectionKey = sk
}
